rust/kernel/net/device.rs

```diff
@@ -0,0 +1,1037 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Net Device Operations.
+//!
+//! C header: [`include/linux/netdevice.h`](../../../../include/linux/netdevice.h)
+
+use core::{marker, mem, ptr};
+
+use crate::bindings;
+use crate::c_from_kernel_result;
+use crate::error::{Error, Result};
+use crate::types::{SavedAsPointer, SavedAsPointerMut};
+use crate::{c_types, CStr};
+
+use super::ethtool::EthToolOps;
+use super::rtnl::{RtnlLinkStats64, RtnlLock};
+use super::skbuff::SkBuff;
+
+extern "C" {
+    #[allow(improper_ctypes)]
+    fn rust_helper_netdev_priv(dev: *const bindings::net_device) -> *mut c_types::c_void;
+
+    #[allow(improper_ctypes)]
+    fn rust_helper_eth_hw_addr_random(dev: *const bindings::net_device);
+
+    #[allow(improper_ctypes)]
+    fn rust_helper_net_device_set_new_lstats(dev: *mut bindings::net_device) -> c_types::c_int;
+
+    #[allow(improper_ctypes)]
+    fn rust_helper_dev_lstats_add(dev: *mut bindings::net_device, len: u32);
+}
+
+/// interface name assignment types (sysfs name_assign_type attribute).
+#[repr(u8)]
+pub enum NetNameAssingType {
+    /// Unknown network name assing type.
+    Unknown = bindings::NET_NAME_UNKNOWN as u8,
+    /// Enum network name assing type.
+    Enum = bindings::NET_NAME_ENUM as u8,
+}
+
+unsafe extern "C" fn setup_netdev_callback<T: NetDeviceAdapter>(dev: *mut bindings::net_device) {
+    let mut dev = NetDevice::<T>::from_pointer(dev);
+
+    T::setup(&mut dev);
+}
+
+/// Wraps the kernel's `struct net_device`.
+///
+/// # Invariants
+///
+/// The pointer [`NetDevice::ptr`] is non-null and valid.
+#[repr(transparent)]
+pub struct NetDevice<T: NetDeviceAdapter> {
+    ptr: *const bindings::net_device,
+    priv_data: marker::PhantomData<T>,
+}
+
+unsafe impl<T: NetDeviceAdapter> Sync for NetDevice<T> {}
+
+impl<T: NetDeviceAdapter> NetDevice<T> {
```

Comment:
Note that Rust's polymorphize support is not yet mature, so one copy of each function might be monomorphized for each `T`. Not a big concern here though.